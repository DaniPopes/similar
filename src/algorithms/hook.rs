use std::convert::Infallible;
use std::mem;

/// A trait for reacting to an edit script from the "old" version to
/// the "new" version.
pub trait DiffHook: Sized {
    type Error;

    /// Called when lines with indices `old_index` (in the old version) and
    /// `new_index` (in the new version) start an section equal in both
    /// versions, of length `len`.
    fn equal(&mut self, old_index: usize, new_index: usize, len: usize) -> Result<(), Self::Error> {
        let _ = old_index;
        let _ = new_index;
        let _ = len;
        Ok(())
    }

    /// Called when a section of length `old_len`, starting at `old_index`,
    /// needs to be deleted from the old version.
    fn delete(
        &mut self,
        old_index: usize,
        old_len: usize,
        new_index: usize,
    ) -> Result<(), Self::Error> {
        let _ = old_index;
        let _ = old_len;
        let _ = new_index;
        Ok(())
    }

    /// Called when a section of the new version, of length `new_len`
    /// and starting at `new_index`, needs to be inserted at position `old_index'.
    fn insert(
        &mut self,
        old_index: usize,
        new_index: usize,
        new_len: usize,
    ) -> Result<(), Self::Error> {
        let _ = old_index;
        let _ = new_index;
        let _ = new_len;
        Ok(())
    }

    /// Called when a section of the old version, starting at index
    /// `old_index` and of length `old_len`, needs to be replaced with a
    /// section of length `new_len`, starting at `new_index`, of the new
    /// version.
    ///
    /// The default implementations invokes `delete` and `insert`.
    fn replace(
        &mut self,
        old_index: usize,
        old_len: usize,
        new_index: usize,
        new_len: usize,
    ) -> Result<(), Self::Error> {
        self.delete(old_index, old_len, new_index)?;
        self.insert(old_index, new_index, new_len)
    }

    /// Always called at the end of the algorithm.
    fn finish(&mut self) -> Result<(), Self::Error> {
        Ok(())
    }
}

impl<'a, D: DiffHook + 'a> DiffHook for &'a mut D {
    type Error = D::Error;

    fn equal(&mut self, old_index: usize, new_index: usize, len: usize) -> Result<(), Self::Error> {
        (*self).equal(old_index, new_index, len)
    }

    fn delete(
        &mut self,
        old_index: usize,
        old_len: usize,
        new_index: usize,
    ) -> Result<(), Self::Error> {
        (*self).delete(old_index, old_len, new_index)
    }

    fn insert(
        &mut self,
        old_index: usize,
        new_index: usize,
        new_len: usize,
    ) -> Result<(), Self::Error> {
        (*self).insert(old_index, new_index, new_len)
    }

    fn replace(
        &mut self,
        old: usize,
        old_len: usize,
        new: usize,
        new_len: usize,
    ) -> Result<(), Self::Error> {
        (*self).replace(old, old_len, new, new_len)
    }

    fn finish(&mut self) -> Result<(), Self::Error> {
        (*self).finish()
    }
}

/// Utility enum to capture a diff operation.
#[derive(Debug, PartialEq, Eq, Hash, Clone, Copy)]
pub enum DiffOp {
    Equal {
        old_index: usize,
        new_index: usize,
        len: usize,
    },
    Delete {
        old_index: usize,
        old_len: usize,
        new_index: usize,
    },
    Insert {
        old_index: usize,
        new_index: usize,
        new_len: usize,
    },
    Replace {
        old_index: usize,
        old_len: usize,
        new_index: usize,
        new_len: usize,
    },
}

/// Captures diff operations.
#[derive(Default, Clone)]
pub struct CaptureHook(Vec<DiffOp>);

impl CaptureHook {
    /// Creates a new capture hook.
    pub fn new() -> CaptureHook {
        CaptureHook::default()
    }

    /// Removes all replace operations.
    pub fn resolve_replace(&mut self) {
        self.0 = mem::replace(&mut self.0, Vec::new())
            .into_iter()
            .filter_map(|op| Some(op))
            .collect();
    }

    /// Converts the capture hook into a vector.
    pub fn into_vec(self) -> Vec<DiffOp> {
        self.0
    }

    /// Accesses the captured operations.
    pub fn ops(&self) -> &[DiffOp] {
        &self.0
    }
}

impl DiffHook for CaptureHook {
    type Error = Infallible;

    fn equal(&mut self, old_index: usize, new_index: usize, len: usize) -> Result<(), Self::Error> {
        self.0.push(DiffOp::Equal {
            old_index,
            new_index,
            len,
        });
        Ok(())
    }

    fn delete(
        &mut self,
        old_index: usize,
        old_len: usize,
        new_index: usize,
    ) -> Result<(), Self::Error> {
        self.0.push(DiffOp::Delete {
            old_index,
            old_len,
            new_index,
        });
        Ok(())
    }

    fn insert(
        &mut self,
        old_index: usize,
        new_index: usize,
        new_len: usize,
    ) -> Result<(), Self::Error> {
        self.0.push(DiffOp::Insert {
            old_index,
            new_index,
            new_len,
        });
        Ok(())
    }

    fn replace(
        &mut self,
        old_index: usize,
        old_len: usize,
        new_index: usize,
        new_len: usize,
    ) -> Result<(), Self::Error> {
        self.0.push(DiffOp::Replace {
            old_index,
            old_len,
            new_index,
            new_len,
        });
        Ok(())
    }
}
