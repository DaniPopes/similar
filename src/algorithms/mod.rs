//! Various diff (longest common subsequence) algorithms.
//!
//! The implementations of the algorithms in this module are relatively low
//! level and expose the most generic bounds possible for the algorithm.  To
//! use them you would typically use the higher level API if possible but
//! direct access to these algorithms can be useful in some cases.
//!
//! All these algorithms provide a `diff` function which takes two indexable
//! objects (for instance slices) and a [`DiffHook`].  As the diff is generated
//! the diff hook is invoked.  Note that the diff hook does not get access to
//! the actual values but only the indexes.  This is why the diff hook is not
//! used outside of the raw algorithm implementations as for most situations
//! access to the values is useful of required.
//!
//! Most of the crate operates on the [`Algorithm`] enum which abstracts over
//! the different algorithms.

// general traits and utilities
mod capture;
mod hook;
mod replace;

use std::hash::Hash;
use std::ops::{Index, Range};

pub use capture::*;
pub use hook::*;
pub use replace::*;

// actual diffing algorithms
pub mod myers;
pub mod patience;

/// An enum representing a diffing algorithm.
#[derive(Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord, Debug)]
pub enum Algorithm {
    Myers,
    Patience,
}

impl Default for Algorithm {
    /// Returns the default algorithm ([`Algorithm::Myers`]).
    fn default() -> Algorithm {
        Algorithm::Myers
    }
}

/// Creates a diff between old and new with the given algorithm.
///
/// Diffs `old`, between indices `old_range` and `new` between indices `new_range`.
pub fn diff<Old, New, D>(
    alg: Algorithm,
    d: &mut D,
    old: &Old,
    old_range: Range<usize>,
    new: &New,
    new_range: Range<usize>,
) -> Result<(), D::Error>
where
    Old: Index<usize> + ?Sized,
    New: Index<usize> + ?Sized,
    D: DiffHook,
    Old::Output: Hash + Eq + Ord,
    New::Output: PartialEq<Old::Output> + Hash + Eq + Ord,
{
    match alg {
        Algorithm::Myers => myers::diff(d, old, old_range, new, new_range),
        Algorithm::Patience => patience::diff(d, old, old_range, new, new_range),
    }
}

/// Shortcut for diffing slices with a specific algorithm.
pub fn diff_slices<D, T>(alg: Algorithm, d: &mut D, old: &[T], new: &[T]) -> Result<(), D::Error>
where
    D: DiffHook,
    T: Eq + Hash + Ord,
{
    diff(alg, d, old, 0..old.len(), new, 0..new.len())
}
