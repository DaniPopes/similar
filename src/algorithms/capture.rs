use crate::algorithms::hook::DiffHook;
use std::convert::Infallible;

/// Utility enum to capture a diff operation.
///
/// This is used by [`Capture`](crate::algorithms::Capture).
#[derive(Debug, PartialEq, Eq, Hash, Clone, Copy)]
pub enum DiffOp {
    /// A segment is equal (see [`DiffHook::equal`])
    Equal {
        old_index: usize,
        new_index: usize,
        len: usize,
    },
    /// A segment was deleted (see [`DiffHook::delete`])
    Delete {
        old_index: usize,
        old_len: usize,
        new_index: usize,
    },
    /// A segment was inserted (see [`DiffHook::insert`])
    Insert {
        old_index: usize,
        new_index: usize,
        new_len: usize,
    },
    /// A segment was replaced (see [`DiffHook::replace`])
    Replace {
        old_index: usize,
        old_len: usize,
        new_index: usize,
        new_len: usize,
    },
}

/// A [`DiffHook`] that captures all diff operations.
#[derive(Default, Clone)]
pub struct Capture(Vec<DiffOp>);

impl Capture {
    /// Creates a new capture hook.
    pub fn new() -> Capture {
        Capture::default()
    }

    /// Converts the capture hook into a vector.
    pub fn into_vec(self) -> Vec<DiffOp> {
        self.0
    }

    /// Accesses the captured operations.
    pub fn ops(&self) -> &[DiffOp] {
        &self.0
    }
}

impl DiffHook for Capture {
    type Error = Infallible;

    fn equal(&mut self, old_index: usize, new_index: usize, len: usize) -> Result<(), Self::Error> {
        self.0.push(DiffOp::Equal {
            old_index,
            new_index,
            len,
        });
        Ok(())
    }

    fn delete(
        &mut self,
        old_index: usize,
        old_len: usize,
        new_index: usize,
    ) -> Result<(), Self::Error> {
        self.0.push(DiffOp::Delete {
            old_index,
            old_len,
            new_index,
        });
        Ok(())
    }

    fn insert(
        &mut self,
        old_index: usize,
        new_index: usize,
        new_len: usize,
    ) -> Result<(), Self::Error> {
        self.0.push(DiffOp::Insert {
            old_index,
            new_index,
            new_len,
        });
        Ok(())
    }

    fn replace(
        &mut self,
        old_index: usize,
        old_len: usize,
        new_index: usize,
        new_len: usize,
    ) -> Result<(), Self::Error> {
        self.0.push(DiffOp::Replace {
            old_index,
            old_len,
            new_index,
            new_len,
        });
        Ok(())
    }
}
