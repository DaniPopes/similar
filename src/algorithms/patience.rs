//! Patience diff algorithm.
//!
//! * time: `O(N log N + M log M + (N+M)D)`
//! * space: `O(N+M)`
//!
//! Tends to give more human-readable outputs. See [Bram Cohen's blog
//! post](https://bramcohen.livejournal.com/73318.html) describing it.
use std::collections::hash_map::Entry;
use std::collections::HashMap;
use std::hash::Hash;
use std::ops::{Index, Range};

use crate::algorithms::{myers, DiffHook, Replace};

struct Indexable<'a, T: ?Sized> {
    value: &'a T,
    index: usize,
}

impl<'a, T: Index<usize> + 'a> std::fmt::Debug for Indexable<'a, T>
where
    T::Output: std::fmt::Debug,
{
    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(fmt, "{:?}", &self.value[self.index])
    }
}

impl<'a, 'b, A, B> PartialEq<Indexable<'a, A>> for Indexable<'b, B>
where
    A: Index<usize> + 'b + ?Sized,
    B: Index<usize> + 'b + ?Sized,
    B::Output: PartialEq<A::Output>,
{
    fn eq(&self, b: &Indexable<'a, A>) -> bool {
        self.value[self.index] == b.value[b.index]
    }
}

fn unique<T>(seq: &T, lower: usize, upper: usize) -> Vec<Indexable<T>>
where
    T: Index<usize> + ?Sized,
    T::Output: Hash + Eq,
{
    let mut by_item = HashMap::new();
    for index in lower..upper {
        match by_item.entry(&seq[index]) {
            Entry::Vacant(entry) => {
                entry.insert(Some(index));
            }
            Entry::Occupied(mut entry) => {
                let entry = entry.get_mut();
                if entry.is_some() {
                    *entry = None
                }
            }
        }
    }
    let mut rv = by_item
        .into_iter()
        .filter_map(|(_, x)| x)
        .map(|index| Indexable { value: seq, index })
        .collect::<Vec<_>>();
    rv.sort_by(|a, b| a.index.cmp(&b.index));
    rv
}

struct Patience<'old, 'new, 'd, Old: ?Sized, New: ?Sized, D> {
    d: &'d mut D,
    old: &'old Old,
    new: &'new New,
    current_old: usize,
    end_old: usize,
    current_new: usize,
    end_new: usize,
    indexes_old: &'old [Indexable<'old, Old>],
    indexes_new: &'new [Indexable<'new, New>],
}

impl<'old, 'new, 'd, Old, New, D> DiffHook for Patience<'old, 'new, 'd, Old, New, D>
where
    D: DiffHook + 'd,
    Old: Index<usize> + ?Sized + 'old,
    New: Index<usize> + ?Sized + 'new,
    New::Output: PartialEq<Old::Output>,
{
    type Error = D::Error;
    fn equal(&mut self, old: usize, new: usize, len: usize) -> Result<(), D::Error> {
        for (old, new) in (old..old + len).zip(new..new + len) {
            let a0 = self.current_old;
            let b0 = self.current_new;
            while self.current_old < self.indexes_old[old].index
                && self.current_new < self.indexes_new[new].index
                && self.new[self.current_new] == self.old[self.current_old]
            {
                self.current_old += 1;
                self.current_new += 1;
            }
            if self.current_old > a0 {
                self.d.equal(a0, b0, self.current_old - a0)?
            }
            myers::diff_offsets(
                self.d,
                self.old,
                self.current_old,
                self.indexes_old[old].index,
                self.new,
                self.current_new,
                self.indexes_new[new].index,
            )?;
            self.current_old = self.indexes_old[old].index;
            self.current_new = self.indexes_new[new].index;
        }
        Ok(())
    }

    fn finish(&mut self) -> Result<(), D::Error> {
        myers::diff(
            self.d,
            self.old,
            self.current_old..self.end_old,
            self.new,
            self.current_new..self.end_new,
        )
    }
}

/// Patience diff algorithm.
///
/// Diff `old`, between indices `old_range` and `new` between indices `new_range`.
pub fn diff<Old, New, D>(
    d: &mut D,
    old: &Old,
    old_range: Range<usize>,
    new: &New,
    new_range: Range<usize>,
) -> Result<(), D::Error>
where
    Old: Index<usize> + ?Sized,
    New: Index<usize> + ?Sized,
    Old::Output: Hash + Eq,
    New::Output: PartialEq<Old::Output> + Hash + Eq,
    D: DiffHook,
{
    let indexes_old = unique(old, old_range.start, old_range.end);
    let indexes_new = unique(new, old_range.start, old_range.end);

    let mut d = Replace::new(Patience {
        d,
        old,
        new,
        current_old: old_range.start,
        end_old: old_range.end,
        current_new: new_range.start,
        end_new: new_range.end,
        indexes_old: &indexes_old,
        indexes_new: &indexes_new,
    });
    myers::diff(
        &mut d,
        &indexes_old,
        0..indexes_old.len(),
        &indexes_new,
        0..indexes_new.len(),
    )?;
    Ok(())
}

/// Shortcut for diffing slices.
pub fn diff_slices<D, T>(d: &mut D, old: &[T], new: &[T]) -> Result<(), D::Error>
where
    D: DiffHook,
    T: Eq + Hash,
{
    diff(d, old, 0..old.len(), new, 0..new.len())
}

#[test]
fn test_patience() {
    let a: &[usize] = &[11, 1, 2, 2, 3, 4, 4, 4, 5, 47, 19];
    let b: &[usize] = &[10, 1, 2, 2, 8, 9, 4, 4, 7, 47, 18];

    struct D(Vec<(usize, usize, usize, usize)>);
    impl DiffHook for D {
        type Error = ();
        fn delete(&mut self, o: usize, len: usize, new: usize) -> Result<(), ()> {
            self.0.push((o, len, new, 0));
            Ok(())
        }
        fn insert(&mut self, o: usize, n: usize, len: usize) -> Result<(), ()> {
            self.0.push((o, 0, n, len));
            Ok(())
        }
        fn replace(&mut self, o: usize, l: usize, n: usize, nl: usize) -> Result<(), ()> {
            self.0.push((o, l, n, nl));
            Ok(())
        }
    }
    let mut d = Replace::new(D(Vec::new()));
    diff(&mut d, a, 0..a.len(), b, 0..b.len()).unwrap();
    let d: D = d.into_inner();

    insta::assert_json_snapshot!(&d.0.as_slice(), @r###"
    [
      [
        0,
        1,
        0,
        1
      ],
      [
        4,
        2,
        4,
        2
      ],
      [
        8,
        1,
        8,
        1
      ],
      [
        10,
        1,
        10,
        1
      ]
    ]
    "###);
}
