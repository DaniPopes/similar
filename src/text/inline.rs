#![cfg(feature = "inline")]
use std::{fmt, iter};

use crate::algorithms::{Algorithm, DiffOp, DiffTag};
use crate::text::{Change, ChangeTag, TextDiff};

use super::split_unicode_words;

use std::ops::Range;

struct MultiIndex<'a, 's> {
    seq: &'a [&'s str],
    value: &'s str,
}

impl<'a, 's> MultiIndex<'a, 's> {
    pub fn new(seq: &'a [&'s str], value: &'s str) -> MultiIndex<'a, 's> {
        MultiIndex { seq, value }
    }

    pub fn get_slice(&self, rng: Range<usize>) -> &'s str {
        let mut start = 0;
        for &sseq in &self.seq[..rng.start] {
            start += sseq.len();
        }
        let mut end = start;
        for &sseq in &self.seq[rng.start..rng.end] {
            end += sseq.len();
        }
        &self.value[start..end]
    }
}

/// Represents the expanded textual change with inline highlights.
///
/// This is like [`Change`] but with inline highlight info.
#[derive(Debug, PartialEq, Eq, Hash, Clone, Ord, PartialOrd)]
pub struct InlineChange<'s> {
    tag: ChangeTag,
    old_index: Option<usize>,
    new_index: Option<usize>,
    values: Vec<(bool, &'s str)>,
    missing_newline: bool,
}

impl<'s> InlineChange<'s> {
    /// Returns the change tag.
    pub fn tag(&self) -> ChangeTag {
        self.tag
    }

    /// Returns the old index if available.
    pub fn old_index(&self) -> Option<usize> {
        self.old_index
    }

    /// Returns the new index if available.
    pub fn new_index(&self) -> Option<usize> {
        self.new_index
    }

    /// Returns the changed values.
    pub fn values(&self) -> &[(bool, &'s str)] {
        &self.values
    }

    /// Returns `true` if this change needs to be followed up by a
    /// missing newline.
    pub fn is_missing_newline(&self) -> bool {
        self.missing_newline
    }
}

impl<'s> From<Change<'s>> for InlineChange<'s> {
    fn from(change: Change<'s>) -> InlineChange<'s> {
        InlineChange {
            tag: change.tag(),
            old_index: change.old_index(),
            new_index: change.new_index(),
            values: vec![(false, change.value())],
            missing_newline: change.missing_newline(),
        }
    }
}

impl<'s> fmt::Display for InlineChange<'s> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        for &(emphasized, value) in &self.values {
            let marker = match (emphasized, self.tag) {
                (false, _) | (true, ChangeTag::Equal) => "",
                (true, ChangeTag::Delete) => "-",
                (true, ChangeTag::Insert) => "+",
            };
            write!(f, "{}{}{}", marker, value, marker)?;
        }
        if self.missing_newline {
            writeln!(f)?;
        }
        Ok(())
    }
}

pub(crate) fn iter_inline_changes<'diff>(
    diff: &'diff TextDiff,
    op: &DiffOp,
) -> impl Iterator<Item = InlineChange<'diff>> {
    let mut change_iter = diff.iter_changes(op).peekable();
    let mut skip_next = false;
    let newline_terminated = diff.newline_terminated;

    iter::from_fn(move || {
        if skip_next {
            change_iter.next();
            skip_next = false;
        }
        if let Some(change) = change_iter.next() {
            let next_change = change_iter.peek();
            match (change.tag, next_change.map(|x| x.tag())) {
                (ChangeTag::Delete, Some(ChangeTag::Insert)) => {
                    let old_value = change.value();
                    let new_value = next_change.unwrap().value();
                    let old_chars = split_unicode_words(&old_value).collect::<Vec<_>>();
                    let new_chars = split_unicode_words(&new_value).collect::<Vec<_>>();
                    let old_mindex = MultiIndex::new(&old_chars, old_value);
                    let new_mindex = MultiIndex::new(&new_chars, new_value);
                    let inline_diff = TextDiff::configure()
                        .algorithm(Algorithm::Patience)
                        .diff_slices(&old_chars, &new_chars);

                    if inline_diff.ratio() < 0.5 {
                        return Some(None.into_iter().chain(Some(change.into()).into_iter()));
                    }

                    // skip the next element as we handle it here
                    skip_next = true;

                    let mut old_values = vec![];
                    let mut new_values = vec![];
                    for op in inline_diff.ops() {
                        match op.tag() {
                            DiffTag::Equal => {
                                old_values.push((false, old_mindex.get_slice(op.old_range())));
                                new_values.push((false, old_mindex.get_slice(op.old_range())));
                            }
                            DiffTag::Delete => {
                                old_values.push((true, old_mindex.get_slice(op.old_range())));
                            }
                            DiffTag::Insert => {
                                new_values.push((true, new_mindex.get_slice(op.new_range())));
                            }
                            DiffTag::Replace => {
                                old_values.push((true, old_mindex.get_slice(op.old_range())));
                                new_values.push((true, new_mindex.get_slice(op.new_range())));
                            }
                        }
                    }

                    Some(
                        Some(InlineChange {
                            tag: ChangeTag::Delete,
                            old_index: change.old_index(),
                            new_index: None,
                            values: old_values,
                            missing_newline: newline_terminated
                                && !old_value.ends_with(&['\r', '\n'][..]),
                        })
                        .into_iter()
                        .chain(
                            Some(InlineChange {
                                tag: ChangeTag::Insert,
                                old_index: None,
                                new_index: next_change.unwrap().new_index(),
                                values: new_values,
                                missing_newline: newline_terminated
                                    && !new_value.ends_with(&['\r', '\n'][..]),
                            })
                            .into_iter(),
                        ),
                    )
                }
                _ => Some(None.into_iter().chain(Some(change.into()).into_iter())),
            }
        } else {
            None
        }
    })
    .flatten()
}

#[test]
fn test_line_ops_inline() {
    let diff = TextDiff::from_lines(
        "Hello World\nsome stuff here\nsome more stuff here\n\nAha stuff here\nand more stuff",
        "Stuff\nHello World\nsome amazing stuff here\nsome more stuff here\n",
    );
    assert_eq!(diff.newline_terminated(), true);
    let changes = diff
        .ops()
        .iter()
        .flat_map(|op| diff.iter_inline_changes(op))
        .collect::<Vec<_>>();
    insta::assert_debug_snapshot!(&changes);
}
