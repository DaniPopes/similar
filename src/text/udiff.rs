//! This module provides unified diff functionality.
//!
//! This module is available for as long as the `text` feature is enabled which
//! is enabled by default.
//!
//! ```rust
//! use similar::text::TextDiff;
//! # let old_text = "";
//! # let new_text = "";
//! let text_diff = TextDiff::from_lines(old_text, new_text);
//! print!("{}", text_diff
//!     .unified_diff()
//!     .context_radius(10)
//!     .header("old_file", "new_file"));
//! ```

use std::fmt;
use std::ops::Range;

use crate::algorithms::{Algorithm, DiffOp};
use crate::text::{Change, ChangeTag, TextDiff};

#[derive(Copy, Clone, Debug)]
struct UnifiedDiffHunkRange(usize, usize);

impl UnifiedDiffHunkRange {
    fn new(range: Range<usize>) -> UnifiedDiffHunkRange {
        UnifiedDiffHunkRange(range.start, range.end)
    }

    fn start(&self) -> usize {
        self.0
    }

    fn end(&self) -> usize {
        self.1
    }
}

impl fmt::Display for UnifiedDiffHunkRange {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let mut beginning = self.start();
        let len = self.end() - self.start();
        if len == 1 {
            write!(f, "{}", beginning)
        } else {
            if len == 0 {
                // empty ranges begin at line just before the range
                beginning -= 1;
            }
            write!(f, "{},{}", beginning, len)
        }
    }
}

/// Unified diff hunk header formatter.
pub struct UnifiedHunkHeader {
    old_range: UnifiedDiffHunkRange,
    new_range: UnifiedDiffHunkRange,
}

impl UnifiedHunkHeader {
    /// Creates a hunk header from a (non empty) slice of diff ops.
    pub fn new(ops: &[DiffOp]) -> UnifiedHunkHeader {
        UnifiedHunkHeader {
            old_range: UnifiedDiffHunkRange::new(ops[0].old_range()),
            new_range: UnifiedDiffHunkRange::new(ops[ops.len() - 1].new_range()),
        }
    }
}

impl fmt::Display for UnifiedHunkHeader {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "@@ -{} +{} @@", &self.old_range, &self.new_range)
    }
}

/// Unified diff formatter.
///
/// The `Display` implementation renders a unified diff.
pub struct UnifiedDiff<'diff, 'old, 'new, 'bufs> {
    diff: &'diff TextDiff<'old, 'new, 'bufs>,
    context_radius: usize,
    header: Option<(String, String)>,
}

impl<'diff, 'old, 'new, 'bufs> UnifiedDiff<'diff, 'old, 'new, 'bufs> {
    /// Creates a formatter from a text diff object.
    pub fn from_text_diff(diff: &'diff TextDiff<'old, 'new, 'bufs>) -> Self {
        UnifiedDiff {
            diff,
            context_radius: 3,
            header: None,
        }
    }

    /// Changes the context radius.
    ///
    /// The context radius is the number of lines between changes that should
    /// be emitted.  This defaults to `3`.
    pub fn context_radius(&mut self, n: usize) -> &mut Self {
        self.context_radius = n;
        self
    }

    /// Sets a header to the diff.
    ///
    /// `a` and `b` are the file names that are added to the top of the unified
    /// file format.  The names are accepted verbaitim which lets you encode
    /// a timestamp into it when separated by a tab (`\t`).  For more information
    /// see [the unified diff format specification](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/diff.html#tag_20_34_10_07)
    pub fn header(&mut self, a: &str, b: &str) -> &mut Self {
        self.header = Some((a.to_string(), b.to_string()));
        self
    }

    /// Iterates over all hunks as configured.
    pub fn iter_hunks(&self) -> impl Iterator<Item = UnifiedDiffHunk<'diff, 'old, 'new, 'bufs>> {
        let diff = self.diff;
        self.diff
            .grouped_ops(self.context_radius)
            .into_iter()
            .filter(|ops| !ops.is_empty())
            .map(move |ops| UnifiedDiffHunk::new(ops, diff))
    }

    fn header_opt(&mut self, header: Option<(&str, &str)>) -> &mut Self {
        if let Some((a, b)) = header {
            self.header(a, b);
        }
        self
    }
}

/// Unified diff hunk formatter.
///
/// The `Display` this renders out a single unified diff's hunk.
pub struct UnifiedDiffHunk<'diff, 'old, 'new, 'bufs> {
    diff: &'diff TextDiff<'old, 'new, 'bufs>,
    ops: Vec<DiffOp>,
}

impl<'diff, 'old, 'new, 'bufs> UnifiedDiffHunk<'diff, 'old, 'new, 'bufs> {
    /// Creates a new hunk for some operations.
    pub fn new(
        ops: Vec<DiffOp>,
        diff: &'diff TextDiff<'old, 'new, 'bufs>,
    ) -> UnifiedDiffHunk<'diff, 'old, 'new, 'bufs> {
        UnifiedDiffHunk { diff, ops }
    }

    /// Returns the header for the hunk.
    pub fn header(&self) -> UnifiedHunkHeader {
        UnifiedHunkHeader::new(&self.ops)
    }

    /// Returns all operations in the hunk.
    pub fn ops(&self) -> &[DiffOp] {
        &self.ops
    }

    /// Iterates over all changes in a hunk.
    pub fn iter_changes(&self) -> impl Iterator<Item = Change<'_>> + '_ {
        // unclear why this needs Box::new here.  It seems to infer some really
        // odd lifetimes I can't figure out how to work with.
        (Box::new(
            self.ops()
                .iter()
                .flat_map(move |op| self.diff.iter_changes(op)),
        )) as Box<dyn Iterator<Item = _>>
    }
}

impl<'diff, 'old, 'new, 'bufs> fmt::Display for UnifiedDiffHunk<'diff, 'old, 'new, 'bufs> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let nl = if self.diff.newline_terminated() {
            ""
        } else {
            "\n"
        };
        writeln!(f, "{}", self.header())?;
        for change in self.iter_changes() {
            write!(
                f,
                "{}{}{}",
                match change.tag() {
                    ChangeTag::Equal => ' ',
                    ChangeTag::Delete => '-',
                    ChangeTag::Insert => '+',
                },
                change.value(),
                nl
            )?;
        }
        Ok(())
    }
}

impl<'diff, 'old, 'new, 'bufs> fmt::Display for UnifiedDiff<'diff, 'old, 'new, 'bufs> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let mut header = self.header.as_ref();
        for hunk in self.iter_hunks() {
            if let Some((old_file, new_file)) = header.take() {
                writeln!(f, "--- {}", old_file)?;
                writeln!(f, "+++ {}", new_file)?;
            }
            write!(f, "{}", hunk)?;
        }
        Ok(())
    }
}

/// Quick way to get a unified diff as string.
///
/// `n` configures [`UnifiedDiff::context_radius`] and
/// `header` configures [`UnifiedDiff::header`] when not `None`.
pub fn unified_diff<'old, 'new>(
    alg: Algorithm,
    old: &'old str,
    new: &'new str,
    n: usize,
    header: Option<(&str, &str)>,
) -> String {
    TextDiff::configure()
        .algorithm(alg)
        .diff_lines(old, new)
        .unified_diff()
        .context_radius(n)
        .header_opt(header)
        .to_string()
}

#[test]
fn test_unified_diff() {
    let diff = TextDiff::from_lines(
        "a\nb\nc\nd\ne\nf\ng\nh\ni\nj\nk\nl\nm\nn\no\np\nq\nr\ns\nt\nu\nv\nw\nx\ny\nz\nA\nB\nC\nD\nE\nF\nG\nH\nI\nJ\nK\nL\nM\nN\nO\nP\nQ\nR\nS\nT\nU\nV\nW\nX\nY\nZ",
        "a\nb\nc\nd\ne\nf\ng\nh\ni\nj\nk\nl\nm\nn\no\np\nq\nr\nS\nt\nu\nv\nw\nx\ny\nz\nA\nB\nC\nD\nE\nF\nG\nH\nI\nJ\nK\nL\nM\nN\no\nP\nQ\nR\nS\nT\nU\nV\nW\nX\nY\nZ",
    );
    insta::assert_snapshot!(&diff.unified_diff().header("a.txt", "b.txt").to_string());
}
