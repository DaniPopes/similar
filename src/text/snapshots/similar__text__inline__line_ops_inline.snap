---
source: src/text/inline.rs
expression: "&changes"
---
[
    InlineChange {
        tag: Insert,
        old_index: None,
        new_index: Some(
            0,
        ),
        values: [
            (
                false,
                "Stuff\n",
            ),
        ],
        missing_newline: false,
    },
    InlineChange {
        tag: Equal,
        old_index: Some(
            0,
        ),
        new_index: Some(
            1,
        ),
        values: [
            (
                false,
                "Hello World\n",
            ),
        ],
        missing_newline: false,
    },
    InlineChange {
        tag: Delete,
        old_index: Some(
            1,
        ),
        new_index: None,
        values: [
            (
                false,
                "some ",
            ),
            (
                false,
                "stuff here\n",
            ),
        ],
        missing_newline: false,
    },
    InlineChange {
        tag: Insert,
        old_index: None,
        new_index: Some(
            2,
        ),
        values: [
            (
                false,
                "some ",
            ),
            (
                true,
                "amazing ",
            ),
            (
                false,
                "stuff here\n",
            ),
        ],
        missing_newline: false,
    },
    InlineChange {
        tag: Equal,
        old_index: Some(
            2,
        ),
        new_index: Some(
            3,
        ),
        values: [
            (
                false,
                "some more stuff here\n",
            ),
        ],
        missing_newline: false,
    },
    InlineChange {
        tag: Delete,
        old_index: Some(
            3,
        ),
        new_index: None,
        values: [
            (
                false,
                "\n",
            ),
        ],
        missing_newline: false,
    },
    InlineChange {
        tag: Delete,
        old_index: Some(
            4,
        ),
        new_index: None,
        values: [
            (
                false,
                "Aha stuff here\n",
            ),
        ],
        missing_newline: false,
    },
    InlineChange {
        tag: Delete,
        old_index: Some(
            5,
        ),
        new_index: None,
        values: [
            (
                false,
                "and more stuff",
            ),
        ],
        missing_newline: true,
    },
]
